Add char range overload to DotDot operator

IISResetMe opened the pull request:
This PR attempts to allow the DotDot operator (..) to generate
character ranges, in turn allowing shorthand syntax like:

    $AtoZ = -join('A'..'Z')

Since PowerShell has no shorthand syntax for char literals, we
test whether the left-hand-side is of type String, and subsequently
attempts to convert both operands to a char (in accordance with 
PowerShell operator overload behavior in general).

NOTE: 
1) This breaks a number of currently supported range operations, such
as implicit numerical conversion of strings:
'1'..'100'

2) This implementation is fairly naive, making no attempts to assert whether the input characters (or any codepoints in between) fall outside UCS-2 (ie. surrogates). Attempts to expand ranges bound by surrogate pairs are expected to fail and not covered by this implementation (I will leave this an exercise for someone more experienced with unicode hackery than myself)

I was unable to locate existing tests for the integer range function, please advise on where to add those.

IISResetMe committed:
Add char range overload to DotDot operator

This commit attempts to allow the DotDot operator (..) to generate
character ranges.

Since PowerShell has no shorthand syntax for char literals, this change
tests whether the left-hand-side is of type String, and subsequently
attempts to convert both operands to a char.

NOTE: This breaks a number of currently supported range operations, such
as implicit numerical conversion of strings:
'1'..'100'

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=5026) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=5026) before we can accept your contribution.<br/><br/>:x: IISResetMe [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=5026)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=5026) it.</sub>

lzybkr commented on src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs:
```diff
@@ -863,5 +863,34 @@ internal static object CompareIne(char lhs, char rhs)
             return firstAsUpper != secondAsUpper ? Boxed.True : Boxed.False;
         }
+
+        internal static object[] Range(char start, char end)
```
What do you think about returning `char[]` instead?

iSazonov replied:
@IISResetMe Could you please address the comment?

IISResetMe replied:
I like the idea, but for now I think we should keep operator output consistent (ie. return `object[]` with individual items of the expected type).

There are are number of builtin operators that return `object[]` in this fashion (`@()`, `-split`, any comparison operator in array/filter mode etc.), changing the behavior should be a concerted effort across all of them IMO

lzybkr commented on src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs:
```diff
@@ -863,5 +863,34 @@ internal static object CompareIne(char lhs, char rhs)
+            int lower = (int)start;
+            int upper = (int)end;
+
+            if (lower == upper)
```
I think this special case is correctly handled by the code below and it happens rarely, so maybe we should remove it. (And I see you copied the code from the `int` version, so you could remove it there too.)

IISResetMe replied:
Mainly did it this way to keep it similar to the existing code in Compiler.cs. I'm inclined to make a single Range method that takes the operands as Expression's, de-duplicate the loop that generates the actual sequence and then cast the entire thing to char[], int[] or long[] based on the lhs operand before returning. Will update the code later today

iSazonov replied:
@IISResetMe Is this addresed?

IISResetMe replied:
I've removed the redundant branch in both overloads now. https://github.com/PowerShell/PowerShell/pull/5026/commits/0406cb37f1556b8497325236672819b54da23e53

lzybkr reviewed (changes requested):
The code changes look fine - but we need tests.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -4901,6 +4903,11 @@ public object VisitBinaryExpression(BinaryExpressionAst binaryExpressionAst)
                     return Expression.Call(CachedReflectionInfo.TypeOps_AsOperator, lhs.Cast(typeof(object)), rhs.Convert(typeof(Type)));
 
                 case TokenKind.DotDot:
+                    if(lhs.Type == typeof(string)){
```
Will this work:
```
'A'..'Z' | % { ... }
```

IISResetMe replied:
Yes, this will produce the sequence A (0x41) through Z (0x5A)

kborowinski replied:
Apparently it *does not work* on the Powershell Core 6.0.0 GA
```powershell
> 'A'..'Z' | % {$_ }
Cannot convert value "A" to type "System.Int32". Error: "Input string was not in a correct format."
At line:1 char:1
+ 'A'..'Z' | % {$_ }
+ ~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (:) [], RuntimeException
+ FullyQualifiedErrorId : InvalidCastFromStringToInteger
```

markekraus replied:
@kborowinski there is an open issue for that #5519

kborowinski replied:
@markekraus: sorry, I've missed that issue

iSazonov commented on src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs:
```diff
@@ -863,5 +863,34 @@ internal static object CompareIne(char lhs, char rhs)
+        internal static object[] Range(char start, char end)
+        {
+            int lower = (int)start;
+            int upper = (int)end;
```
I wonder - is this ASCII Arithmetic work correctly for unicode?

lzybkr replied:
It won't work with surrogate pairs because those can't be represented in a `char`, but I don't think that is an interesting scenario.

iSazonov replied:
Since we added `u{} should we detect surrogate pairs and write an error?

lzybkr replied:
I guess we should error on invalid characters (half of the pair). We could also support surrogate pair ranges if we didn't convert the string to char immediately, instead analyzing the string more thoroughly. But I'm not sure I'd block this PR on either of those.

iSazonov replied:
I guess it will work with traditional languages in common scenarios but before I saw issues from China and Japan mans - if the new operator does not matter to them I agree with you.

iSazonov replied:
@IISResetMe Could you please address the comment in code or add relevant comments in the PR description for future documentation?

IISResetMe replied:
@iSazonov I don't feel confident enough in my unicode-fu to attempt to implement the appropriate handling in code for this at the moment, I've updated the PR description to reflect this deficiency instead.

iSazonov replied:
Thanks. 

Closed.

IISResetMe committed:
Add RangeOperator.Tests.ps1

This commit adds basic tests for the current integer range (..) operator

IISResetMe commented:
Oops, rebased powershell:master into the branch, looks like all parent commits are in the PR now, how do I clean it up? 😞

markekraus commented:
@IISResetMe did you only have the 2 commits? if so you could do this (assuming `PowerShell` is the name of your remote branch for the actual PowerShell repo). Make sure you stash or commit any workspace changes before doing this as they will be lost. I'm not a git expert, and maybe there is a better way. So maybe wait until someone more seasoned either "thumbs up" this or tells you just how wrong I am :)


```bash
git fetch
git checkout -b 20171015fix PowerShell/master
git cherry-pick 6e6dfffc2d0150b504585597ad4df997a174867a
git cherry-pick 6d6f20d28157ee49071f018051ac7b35be57bbb0
git checkout -b 20171015backup overload-DotDot-CharRange-1
git checkout overload-DotDot-CharRange-1
git reset --hard 20171015fix
git push --force
```

This will create a new "20171015fix" branch at the current master on this repo, then cherry-picks your 2 commits, then backs up your existing branch to `20171015backup`, and then hard reset your branch to the "20171015fix" branch. This will both "rebase" your branch on to master and add your commits with out the merge mess.  This will likely mess up any comment history here, but I guess it's already some what messed up.

IISResetMe reviewed (commented):
These were the most basic tests I could think of for the existing Range operator. All examples I could find in the about_* help files pass these.

IISResetMe commented:
Thanks @markekraus, that seems to have done the trick.

IISResetMe committed:
Change RangeOperator.Tests.ps1 encoding, add tests

Add tests for the [char] overload feature. RangeOperator.Tests.ps1 encoding changed to UTF16LE (Windows "Unicode") to allows inline unicode chars in test case

IISResetMe committed:
Change RangeOperator.Tests.ps1 encoding to UTF-8

Changed RangeOperator.Tests.ps1 encoding to UTF-8 (no BOM), substituted UTF-16 char literals with [char] casts

IISResetMe commented:
@TravisEz13 Done, https://github.com/IISResetMe/PowerShell/blob/34bc59046698d0de41bb5403a6105c7070b38a46/test/powershell/Language/Operators/RangeOperator.Tests.ps1 is UTF-8 (with updated tests) now

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+        It "Range operator generates arrays of integers" {
+            $Range = 5..8
+            $Range.count | Should Be 4
+            $Range[0] | Should BeOfType [int]
```
Please add tests for values - `$Range[0] | Should Be 5`

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+        It "Range operator accepts negative integer values" {
+            {
+                -8..-5
+            } |Should Not Throw
```
Please remove the should - next line do the same and Pester can catch a throw there correctly.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+
+            $Range = -8..-5            
+            $Range.count | Should Be 4
+            $Range[0] |Should Be -8
```
Formatting - please add space after |

Please add type check as in previous test.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+        It "Range operator support single-item sequences" {
+            {
+                0..0
+            } |Should Not Throw
```
Please remove.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            $Range[0] | Should Be 0
+        }
+
+        It "Range operator works in ascending and descending order" {
```
Previous tests already test "ascending". I think we could test here only descending order for positive and negative numbers. But I believe we should replace some tests with one - "-2 .. 2" and "2 .. -2" - so we check transition through zero too.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    Context "Character expansion" {
+
```
Please remove extra line.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+
+    Context "Character expansion" {
+
+        It "Range operator generates an array of [char] from single-character string operands" {
```
I think we can remove the test and add type checks in next test.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+        It "Range operator works with 16-bit unicode characters" {
+            {
+                "$([char]0x0110)".."$([char]0x0114)"
+            } | Should Not Throw
```
Please remove.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            {
+                "$([char]0x0110)".."$([char]0x0114)"
+            } | Should Not Throw
+            $UnicodeRange = "$([char]0x0110)".."$([char]0x0114)"
```
Please use new syntax `u{0x0110}

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            } | Should Not Throw
+            $UnicodeRange = "$([char]0x0110)".."$([char]0x0114)"
+            $UnicodeRange.count | Should Be 5
+            $UnicodeRange.Where({$_ -is [char]}).count | Should Be 5
```
Please check values too as above for numbers.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    Context "Range operator operand types" {
+
```
Please remove extra line.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+
+    Context "Range operator operand types" {
+
+        It "Range operator works on [int]" {
```
Please remove the test - we do the checks above already.

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            } | Should Not Throw
+            $Range = ([long]1)..([long]10)
+            $Range.count | Should Be 10
+            $Range.Where({$_ -is [int]}).count | Should Be 10
```
Is it really [int]? I think we should use values > int.MaxValue.
And please check values too.

IISResetMe replied:
Current implementation doesn't support values > int.MaxValue, intended to add this in a separate branch/PR.

iSazonov replied:
Thanks for clarify. Please remove the confuse tests.

iSazonov replied:
@IISResetMe Please address the comment about the tests for [long] and [bigint].
We should remove its - they don't make sense.

IISResetMe replied:
You're right, they don't make too much sense at the moment. I've removed them now

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            } | Should Not Throw
+            $Range = ([bigint]1)..([bigint]10)
+            $Range.count | Should Be 10
+            $Range.Where({$_ -is [int]}).count | Should Be 10
```
Is it really [int]? I think we should use values > long.MaxValue.
And please check values too.

IISResetMe replied:
The existing numeric range operator only supports int regardless of input type

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            {
+                1.2d..2.9d
+            } | Should Not Throw
+            $Range = 1.1d..9.9d
```
Why so much? I think enough `$Range = 1.1d..3.9d`

iSazonov commented on test/powershell/Language/Operators/RangeOperator.Tests.ps1:
```diff
@@ -0,0 +1,135 @@
+            $Range[9] | Should Be 10
+        }
+    }
+}
```
Please add new line at EOF.

IISResetMe committed:
Update RangeOperator.Tests.ps1

Fixed formatting, removed unnecessary checks

iSazonov commented:
@TravisEz13 I updated three comment above - after they have been resolved we'll ready to merge.

IISResetMe committed:
Remove redundant code from Range()

Both Range() overloads have unnecessary branches for the special case of lhs == rhs, removed for simplicity.

IISResetMe committed:
Remove confusing/unnecessary RangeOp tests

Removed tests for .. using [long] and [bigint] input operands. I believe these will make more sense once we add support for ranges with values outside Int32, but for now might be confusing

IISResetMe commented:
@iSazonov I believe I've addressed all outstanding comments :-)

iSazonov commented:
@IISResetMe Many thanks! We are near to merge. Please address @lzybkr comment above.

IISResetMe commented:
@iSazonov Done

iSazonov commented:
@lzybkr Could you please take another look if you have a bit free time?
